services/metadata_service/api/tag.py

```diff
@@ -0,0 +1,139 @@
+                  enum: [add, remove]
+                tag:
+                  type: string
+                user:
+                  type: string
+        produces:
+        - application/json
+        responses:
+            "202":
+                description: successful operation. Return newly registered task
+            "404":
+                description: not found
+            "500":
+                description: internal server error
+        """
+        body = await request.json()
+        results = []
+        for o in body:
+            try:
+                table = self._get_table(o['object_type'])
+                pathspec = o['id'].split('/')
+                # Do some basic verification
+                if o['object_type'] == 'flow' and len(pathspec) != 1:
+                    raise ValueError("invalid flow specification: %s" % o['id'])
+                elif o['object_type'] == 'run' and len(pathspec) != 2:
+                    raise ValueError("invalid run specification: %s" % o['id'])
+                elif o['object_type'] == 'step' and len(pathspec) != 3:
+                    raise ValueError("invalid step specification: %s" % o['id'])
+                elif o['object_type'] == 'task' and len(pathspec) != 4:
+                    raise ValueError("invalid task specification: %s" % o['id'])
+                elif o['object_type'] == 'artifact' and len(pathspec) != 5:
+                    raise ValueError("invalid artifact specification: %s" % o['id'])
+                obj_filter = table.get_filter_dict(*pathspec)
+            except ValueError as e:
+                return DBResponse(response_code=400, body=json.dumps(
+                    {"message": "invalid input: %s" % str(e)}))
+
+            # Now we can get the object
+            obj = await table.get_records(
+                filter_dict=obj_filter, fetch_single=True, expanded=True)
+            if obj.response_code != 200:
+                return DBResponse(response_code=obj.response_code, body=json.dumps(
+                    {"message": "could not get object %s: %s" % (o['id'], obj.body)}))
+
+            # At this point do some checks and update the tags
+            obj = obj.body
+            modified = False
+            if o['operation'] == 'add':
+                # This is the only error we fail hard on; adding a tag that is
+                # in system tag
+                if o['tag'] in obj['system_tags']:
+                    return DBResponse(response_code=405, body=json.dumps(
+                        {"message": "tag %s is already a system tag and can't be added to %s"
+                        % (o['tag'], o['id'])}))
+                if o['tag'] not in obj['tags']:
+                    modified = True
+                    obj['tags'].append(o['tag'])
+            elif o['operation'] == 'remove':
+                if o['tag'] in obj['tags']:
+                    modified = True
+                    obj['tags'].remove(o['tag'])
```

Comment:
Noticed that the Metaflow client and service do not guard against duplicate tags, so `--tag test --tag test` would add the test tag twice, due to tags being a simple array. `remove` will only get rid of one occurrence of a tag. Should this be the way the update works?

alternatively get rid of all occurences of a tag:
```
obj['tags'] = [tag for tag in obj['tags] if tag != o['tag']]
```

Reply:
You are right again on the creation side of things; the update will guard against duplicates but not the add (I will double check on that side; I think we should just convert it to a set and set that and also check against system tags).